### PR TITLE
#146 copying transitive dependencies functionality for containerize-d…

### PR DESCRIPTION
**Pull request by Cho-William**
…ependencies goal

**Release Notes Addition Request**
A new non-default execution goal `containerize-dependencies` has been added the habushu-maven-plugin. This execution is helpful for establishing a virtual environment (of a containerized setting) housing monorepo python dependencies in a more resilient pattern. The goal examines the project's habushu-type dependencies declared in the building project's `pom.xml` and copies relevant source files to the build directory. Transitive habushu-type dependencies are also copied over, with the original file structure preserved to maintain continuity for any path-based dependencies specified throughout the monorepo.

### CODE REVIEW

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Stages the source files of a monorepo dependency
+ * to the target directory along with the source file of
+ * any transitive path-based dependencies.
+ */
+
+@Mojo(name = "containerize-dependencies", defaultPhase = LifecyclePhase.VALIDATE)
```
Comment (d-ryan-ashcraft) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: S: seems like this should be in the `package` phase instead since that would put it just-in-time for usage and seems more functionally aligned with what is happening?

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+public class ContainerizeDependenciesMojo extends AbstractHabushuMojo {
+
+    private static final Logger logger = LoggerFactory.getLogger(ContainerizeDependenciesMojo.class);
+
+    @Component
+    protected MavenSession session;
+
+    protected File anchorSourceDirectory;
```
Comment (d-ryan-ashcraft) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: S: seems like it might be useful for these `anchor*` variables to be plugin parameters so they could be configured via the plugin in non-standard cases

File: README.md [hunk 3]
```diff
@@ -162,6 +162,44 @@ For example, developers may use this feature to bind a Habushu module's `compile
 	</plugin>
 ```
 
+### Leveraging the containerize-dependencies Goal to Prepare a Containerized Virtual Environment ###
+If the execution is specified (does not run by default), the `containerize-dependencies` goal will collect the monorepo python dependency specified in the dependencies block of the project's `pom.xml`, including all transitive dependencies. After collecting the set of necessary dependencies, Habushu will copy the source files to the build directory, while preserving the original structure of the dependency modules to ensure that any path-based dependencies can be leveraged as-is.
```
Comment (carter-cundiff) [inline README.md]: `will collect the monorepo python dependency specified` -> ` will collect any monorepo python dependencies specified`

File: README.md [hunk 4]
```diff
@@ -162,6 +162,44 @@ For example, developers may use this feature to bind a Habushu module's `compile
...
+        <dependency>
+          <groupId>your-group-id</groupId>
+          <artifactId>your-artifact-id</artifactId>
+          <version>your-version</version>
+          <type>habushu</type>
+        </dependency>
+    </dependencies>
+```
+
+Any transitive monorepo dependency `pom.xml` files shoud also use this convention to ensure they are captured by the plugin.
```
Comment (carter-cundiff) [inline README.md]: Any transitive monorepo dependency `pom.xml` files -> Any transitive monorepo dependencies

File: README.md (hunk 3 above)
Comment (epwilkins) [inline README.md]: A:  Presumably the latter portion of this instruction becomes simplified after the Dockerfile changes are introduces to focus less on the implementation and more on the outcome.
```suggestion
If the execution is specified (does not run by default), the `containerize-dependencies` goal will collect a single `habushu` dependency specified the project's `pom.xml`, including all transitive  `habushu` dependencies. After collecting the set of necessary dependencies, Habushu will copy the project files to the build directory under `venv-support`, while preserving the original structure of the dependency modules to ensure that any path-based dependencies can be leveraged as-is. This directory can then be copied onto a Docker container and used to create a virtual environment capable of running the target Habushu project.
```

Comment (epwilkins) [reply README.md]: I think v1 of this feature is only intended to support a single Habushu project being containerized, so even if the current code _could_ copy multiple, the second half of this feature (currently in flight) will depend on there being only one.

File: README.md [hunk 5]
```diff
@@ -162,6 +162,44 @@ For example, developers may use this feature to bind a Habushu module's `compile
...
 ```
 
+### Leveraging the containerize-dependencies Goal to Prepare a Containerized Virtual Environment ###
+If the execution is specified (does not run by default), the `containerize-dependencies` goal will collect the monorepo python dependency specified in the dependencies block of the project's `pom.xml`, including all transitive dependencies. After collecting the set of necessary dependencies, Habushu will copy the source files to the build directory, while preserving the original structure of the dependency modules to ensure that any path-based dependencies can be leveraged as-is.
+
+```xml
+	<plugin>
+		<groupId>org.technologybrewery.habushu</groupId>
+		<artifactId>habushu-maven-plugin</artifactId>
+		<extensions>true</extensions>
```
Comment (epwilkins) [inline README.md]: S: I don't think this is necessary for the specific goal at hand. This would be needed for a `habushu` packaged project, but not one simply leveraging this feature.

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        return fileSet;
+    }
+
+    /**
+     * Checks listed habushu-type dependencies against the set of projects included in the Maven build's session
+     * @return the corresponding Maven projects that match the habushu-type dependencies
+     */
+    protected Set<MavenProject> getHabushuProjects() {
+        HashSet<MavenProject> habushuProjects = new HashSet<>();
+        collectHabushuDependenciesAsProjects(project, habushuProjects);
```
Comment (carter-cundiff) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java:115]: Can drop the `project` parameter here and just reference it as `this.project` in the `collectHabushuDependenciesAsProjects` implementation.

File: habushu-maven-plugin/pom.xml [hunk 7]
```diff
@@ -191,6 +209,9 @@
                             cucumber.junit-platform.naming-strategy=long
                         </configurationParameters>
                     </properties>
+                    <excludes>
+                        <exclude>**/ContainerizeDependenciesMojoTestCase.java</exclude>
```
Comment (epwilkins) [inline habushu-maven-plugin/pom.xml]: Q: Why is this necessary for this class, but not the extremely similar `MojoTestCaseWrapper` in fermenter?

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+        return habushuProjects;
+    }
+
+    /**
+     * Collects the projects of the habushu-type dependencies and adds them to the given project set
```
Comment (carter-cundiff) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: Would it be correct to say `Collects the projects with habushu-type dependencies`? Current wording is sort of confusing to me.

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+     * @param habushuProjects the set to append to with matching projects
+     */
+    private void collectHabushuDependenciesAsProjects(MavenProject currentProject, Set<MavenProject> habushuProjects) {
+        Set<String> habushuDeps = currentProject.getDependencies().stream()
+                .filter(d -> HABUSHU.equals(d.getType()))
+                .map(ContainerizeDependenciesMojo::toGav)
+                .collect(Collectors.toSet());
+        for (MavenProject project : getSession().getProjects()) {
+            if (habushuDeps.contains(toGav(project))) {
+                logger.info("Found habushu-type dependency's project {}.", project);
```
Comment (carter-cundiff) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: `Found habushu-type dependency's project {}.` -> `Found habushu-type dependency's within project {}.`

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 8 above)
Comment (Cho-William) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: that's a better wording, can update

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+    protected void doExecute() throws MojoExecutionException, MojoFailureException {
+        if (this.anchorSourceDirectory == null) {
+            this.anchorSourceDirectory = new File(session.getExecutionRootDirectory());
+        }
+
+        Set<MavenProject> habushuProjects = getHabushuProjects();
+        try {
+            copyRelevant(habushuProjects);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
Comment (d-ryan-ashcraft) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: A (Administrative): It is generally bad practice to throw `RuntimeException`, but we have `HabushuException` for just this purpose.  

https://rules.sonarsource.com/java/tag/error-handling/RSPEC-112/?search=RuntimeException

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+
+    @Override
+    protected void doExecute() throws MojoExecutionException, MojoFailureException {
+        if (this.anchorSourceDirectory == null) {
+            this.anchorSourceDirectory = new File(session.getExecutionRootDirectory());
+        }
+
+        Set<MavenProject> habushuProjects = getHabushuProjects();
+        try {
+            copyRelevant(habushuProjects);
```
Comment (epwilkins) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: A: Name could be better. We're not doing any check for relevancy, right? Unless I'm misunderstanding what we mean by "relevant".  Maybe just `copySourceCode`

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+     */
+    @Override
+    public void execute() throws MojoExecutionException, MojoFailureException {
+        doExecute();
+    }
+
+    @Override
+    protected void doExecute() throws MojoExecutionException, MojoFailureException {
+        if (this.anchorSourceDirectory == null) {
+            this.anchorSourceDirectory = new File(session.getExecutionRootDirectory());
```
Comment (epwilkins) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: A: Why make this a field on the class? Same with `anchorOutputDirectory`.  I would think we could just pass it into the method call.

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+    /**
+     * Copies the relevant source files by leveraging {@link FileSet}s to filter appropriately.
+     * @param habushuProjects corresponding projects of the pom's habushu-type dependencies
+     * @throws IOException
+     */
+    private void copyRelevant(Set<MavenProject> habushuProjects) throws IOException {
+        this.anchorOutputDirectory = new File(project.getBuild().getDirectory()
+                + "/venv-support/" + this.anchorSourceDirectory.getName());
+        Path srcRoot = this.anchorSourceDirectory.toPath();
+        Path destRoot = this.anchorOutputDirectory.toPath();
```
Comment (epwilkins) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: A: If we're only using these `Path` objects, seems like we should just construct them directly instead of constructing a `File` first and converting to `Path`.

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 6 above)
Comment (epwilkins) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: A: Might be nice to detect if multiple Habushu dependencies are directly listed. Either log a message and indicate which dependency we're choosing as the "main" dependency for which to build the virtual env, or throw an exception.

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+    }
+    private static String toGav(MavenProject project) {
+        return project.getGroupId() + ":" + project.getArtifactId() + ":" + project.getVersion();
+    }
+
+    protected MavenSession getSession() {
+        return this.session;
+    }
+
+    protected void setAnchorSourceDirectory(File anchorSourceDirectory) {
```
Comment (epwilkins) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: Q: Is this required since we're setting the base directory in the tests??

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    /**
+     * Copies the relevant source files by leveraging {@link FileSet}s to filter appropriately.
+     * @param habushuProjects corresponding projects of the pom's habushu-type dependencies
+     * @throws IOException
+     */
+    private void copyRelevant(Set<MavenProject> habushuProjects) throws IOException {
+        this.anchorOutputDirectory = new File(project.getBuild().getDirectory()
+                + "/venv-support/" + this.anchorSourceDirectory.getName());
```
Comment (epwilkins) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: A: Seems pointless to include the source directory root here as part of the output path. We could just put the files directly under `venv-support` since it's already acting as the "container" for the venv source files.  That would then allow us to move this logic to the `getAnchorOutputDirectory` method.

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+            Path relativeProjectPath = srcRoot.relativize(projectPath);
+            FileSet fileSet = getDefaultFileSet();
+            fileSet.setDirectory(projectPath.toString());
+            dependencyFileSets.put(relativeProjectPath, fileSet);
+        }
+
+        FileSetManager fileSetManager = new FileSetManager();
+        for (Path project : dependencyFileSets.keySet()) {
+            FileSet fileSet = dependencyFileSets.get(project);
+            logger.info("Copying {} files from {}",
```
Comment (d-ryan-ashcraft) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: A: Maybe something like `Staging monorepo dependency files {} from {}` would be a bit more descriptive?

**Review by epwilkins — APPROVED** [review]

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+                Path relativePath = project.resolve(includedFile);
+                Files.createDirectories(destRoot.resolve(relativePath).getParent());
+                Files.copy(srcRoot.resolve(relativePath), destRoot.resolve(relativePath));
+            }
+        }
+    }
+
+    private FileSet getDefaultFileSet() {
+        FileSet fileSet = new FileSet();
+        fileSet.addExclude("target/**");
```
Comment (d-ryan-ashcraft) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: S: `target`, `dist`, and `.venv` are all possible to override or be different.  We need to dynamically get the values.

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+        fileSet.addExclude("dist/**");
+        return fileSet;
+    }
+
+    /**
+     * Checks listed habushu-type dependencies against the set of projects included in the Maven build's session
+     * @return the corresponding Maven projects that match the habushu-type dependencies
+     */
+    protected Set<MavenProject> getHabushuProjects() {
+        HashSet<MavenProject> habushuProjects = new HashSet<>();
```
Comment (d-ryan-ashcraft) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: A: very minor, but should be `Set<MavenProject> habushuProjects = new HashSet<>()`.  Always want it to be the interface on the LHS unless there is specific implementation class functionality you need to use.

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Copies the relevant source files by leveraging {@link FileSet}s to filter appropriately.
+     * @param habushuProjects corresponding projects of the pom's habushu-type dependencies
+     * @throws IOException
+     */
+    private void copyRelevant(Set<MavenProject> habushuProjects) throws IOException {
```
Comment (d-ryan-ashcraft) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: S: in general, not sure I follow what is `private` vs. `protected`.  Recommend making these protected to allow for extension unless there is a good reason not to.

File: habushu-maven-plugin/src/test/java/org/technologybrewery/habushu/ContainerizeDependenciesMojoTestCase.java [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+
+    /**
+     * Overrides super's {@link #newMavenSession(MavenProject)} to delegate to
+     * the new {@link #newDefaultMavenSession()} introduced in {@link ContainerizeDependenciesMojoTestCase},
+     * which sets the defaults that are normally expected by Maven.
+     */
+    @Override
+    protected MavenSession newMavenSession(MavenProject project) {
+        MavenSession session = newDefaultMavenSession();
+        ArrayList<MavenProject> mavenProjects = new ArrayList<>();
```
Comment (d-ryan-ashcraft) [inline habushu-maven-plugin/src/test/java/org/technologybrewery/habushu/ContainerizeDependenciesMojoTestCase.java]: A: see prior comment

File: habushu-maven-plugin/src/test/java/org/technologybrewery/habushu/ContainerizeDependenciesMojoTestCase.java [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+
+        try {
+            ProjectBuilder projectBuilder = lookup(ProjectBuilder.class);
+
+            for(File mavenProjectFile: this.mavenProjectFiles) {
+                mavenProjects.add(projectBuilder.build(mavenProjectFile, session.getProjectBuildingRequest()).getProject());
+            }
+        } catch (Exception e) {
+            throw new RuntimeException(e);
```
Comment (d-ryan-ashcraft) [inline habushu-maven-plugin/src/test/java/org/technologybrewery/habushu/ContainerizeDependenciesMojoTestCase.java]: A: see prior comment

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 9 above)
Comment (carter-cundiff) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: Or should it be `Found project for habushu-type dependency`?

File: habushu-maven-plugin/src/test/resources/specifications/containerize-dependencies.feature [hunk 22]
```diff
@@ -0,0 +1,12 @@
+@containerizeDependencies
+Feature: Test staging source files of specified dependencies and transitive path-based dependencies
+
+  Scenario: One Habushu-type dependency is specified with the default configuration
+    Given a single Habushu-type dependency
```
Comment (d-ryan-ashcraft) [inline habushu-maven-plugin/src/test/resources/specifications/containerize-dependencies.feature]: S: I think language here is very important to make it crisp.  I think we are saying:
```gherkin
Given a single dependency with a packaging type of habushu
...
Then the source files of the dependency and its associated transitive Habushu packaged dependencies are staged for containerization
```

This directly ties it to the both the original pom where it is `<packaging>habushu</packaging>` as well as the dependency where it is `<type>habushu</type>`.

File: README.md (hunk 3 above)
Comment (d-ryan-ashcraft) [inline README.md]: S (Substantive):`including all transitive dependencies` --> `including all Habushu-packaged transitive dependencies`

File: habushu-maven-plugin/pom.xml [hunk 23]
```diff
@@ -191,6 +209,9 @@
                             cucumber.junit-platform.naming-strategy=long
                         </configurationParameters>
                     </properties>
+                    <excludes>
```
Comment (d-ryan-ashcraft) [inline habushu-maven-plugin/pom.xml]: S: I believe if you rename this class to not end with `TestCase`, you can eliminate this.  To me, that's probably worth the simplification.

**Review by d-ryan-ashcraft — COMMENTED** [review]
Looks good overall!  Left some items that I think we should consider prior to merging.

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 14 above)
Comment (carter-cundiff) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: This was originally done for testing to bypass our method of determining the project root directory which recursively called `project.getParent()`. The `project.getParent()` doesn't work with the maven plugin testing harness as the parents cannot be set. However, now that we're using `session.getExecutionRootDirectory()` for determining the root directory, it's possible this bypass is no longer needed.

Comment (carter-cundiff) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: Also possible it might still be necessary though depending on the output of `session.getExecutionRootDirectory()`

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 10 above)
Comment (Cho-William) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: I was supposed to fix this, thank you for the suggest

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 19 above)
Comment (Cho-William) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: this was an oopsie on me, can fix

File: README.md (hunk 3 above)
Comment (Cho-William) [reply README.md]: What Emily said was my thinking. Once we implement the other cases, this README will be refined more

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 9 above)
Comment (Cho-William) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: second one is better to me, updated

File: README.md (hunk 3 above)
Comment (Cho-William) [reply README.md]: incorporated, thank you

File: README.md (hunk 5 above)
Comment (Cho-William) [reply README.md]: that's correct, removed

File: habushu-maven-plugin/pom.xml (hunk 7 above)
Comment (Cho-William) [reply habushu-maven-plugin/pom.xml]: per Ryan's comment, it's due to the naming of the class. this has been fixed

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 12 above)
Comment (Cho-William) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: The anchor attributes were important for setting up the tests. Tried pulling it out, but ended up breaking the tests. For the sake of time, I'm going to keep this as is, but when we clean up the test case in the follow-on, I think we can implement what you're suggesting

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 11 above)
Comment (Cho-William) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: agreed, refactored

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 13 above)
Comment (Cho-William) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: addressed, see changes throughout the mojo and test steps

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 6 above)
Comment (Cho-William) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: I have it as throwing a Habushu exception if the direct dependencies has more than one habushu-type dependency.

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 14 above)
Comment (Cho-William) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: Per previous comment, this is necessary for setting up the tests. Have made a note to look into this for follow-on.

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 15 above)
Comment (Cho-William) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: This `venv-support` was nice to have, because we will eventually be adding the aggregator poetry project under it. Going to keep this for now, but will revisit this for the follow-on.

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 1 above)
Comment (Cho-William) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: agreed, fixed

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 2 above)
Comment (Cho-William) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: You're actually touching on something we intend to implement. This will be a configurable option in the follow-on.

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 16 above)
Comment (Cho-William) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: Agreed, updated

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 17 above)
Comment (Cho-William) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: That will be covered in a follow-on to enable the user to pass in via configs. We just have it this way as the default, or else using it in the baseline will take a very long time, due to `.venv` in particular taking a while to copy

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 18 above)
Comment (Cho-William) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: Absolutely, thank you for the catch

File: habushu-maven-plugin/src/test/resources/specifications/containerize-dependencies.feature (hunk 22 above)
Comment (Cho-William) [reply habushu-maven-plugin/src/test/resources/specifications/containerize-dependencies.feature]: updated, thank you for the suggest!

File: habushu-maven-plugin/pom.xml (hunk 23 above)
Comment (Cho-William) [reply habushu-maven-plugin/pom.xml]: could not figure out for the life of me why this was getting triggered. that worked, thank you!

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 17 above)
Comment (d-ryan-ashcraft) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: Not sure I follow or agree.  You can get the `target`/`dist` directory by adding the appropriate member variable to the Mojo (worth noting that I'm not sure target should be included either):
```java
    /**
     * Directory in which Poetry places generated source and wheel archive
     * distributions.
     */
    @Parameter(defaultValue = "${project.basedir}/dist", readonly = true, required = true)
    protected File distDirectory;

    /**
     * Directory in which Maven places build-time artifacts - should NOT include dist items.
     */
    @Parameter(defaultValue = "${project.basedir}/target", readonly = true, required = true)
    protected File targetDirectory;

```

Similarly, we can get the virtual environment with an existing call if you look at `CleanHabushuMojo.findCurrentVirtualEnvironmentFullPath()`

These seem like low hanging fruit and important.  I'm a non-concur on pushing them to the next ticket without some additional rationale.

File: README.md (hunk 3 above)
Comment (Cho-William) [reply README.md]: fixed

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 2 above)
Comment (d-ryan-ashcraft) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: This is also a super easy change - if you just add something similar to the following:
```java
@Parameter(defaultValue = "${project.basedir}/venv-support", readonly = true, required = true)
```

As such, we should do it here versus kick it to the next ticket IMO

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 17 above)
Comment (Cho-William) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: Suggestion has been implemented. Please note that `findCurrentVirtualEnvironmentFullPath()` and `getCleanVirtualEnvironmentPath()` have been refactored to live in `HabushuUtil`. `findCurrentVirtualEnvironmentFullPath()` now accepts (as arguments) the various parameters it got from the `CleanHabushuMojo`. Necessary parameters have also been added to the `ContainerizeDepsMojo` as well so that it can call on `findCurrentVirtualEnvironmentFullPath()` appropriately.

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 2 above)
Comment (Cho-William) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: The `venv-support` directory has been refactored to `containerization-support`, and is now a configurable parameter. This effectively parameterizes the `anchorOutputDirectory` attribute.  

`anchorSourceDirectory` is also now a configurable parameter. Note that `session.getExecutionRootDirectory()` is not callable for the parameter's default value in the `@Parameter()` annotation, so it is populated with the default value at the start of the mojo's execution.

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 12 above)
Comment (Cho-William) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: Update: these are now configurable parameters

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java (hunk 13 above)
Comment (Cho-William) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/ContainerizeDependenciesMojo.java]: Update: these were changed back to `File` objects because mojos cannot create `Path` objects via the `@Parameter()` annotation. (This is a result of making the anchor attributes configurable mojo parameters per other feedback)

**Comment by Cho-William** [comment]
PR feedback has been addressed, along with updated documentation to reflect the available configurations.

**Review by d-ryan-ashcraft — APPROVED** [review]

File: README.md [hunk 24]
```diff
@@ -847,6 +891,42 @@ Controls whether the build will continue if lint identifies code that violate ch
...
+
+#### anchorSourceDirectory ####
+
+Controls the directory of where containerization files will be copied from with the `containerize-dependencies` goal. If set, this value should be a directory path that houses all the necessary monorepo dependency Habushu modules (including transitive monorepo dependency modules). Additionally, if set, this path should either be an absolute path or a path relative to the `${project.basedir}` of the `containerize-dependencies` goal's execution. Typically, it is not recommended to set this parameter, as the default directory will be sufficient for monorepo use-cases.
+
+Default: the build's execution root directory
+
+#### containerizeSupportDirectory ####
+
+Controls the location of where containerization files will be placed to with the `containerize-dependencies` goal.
```
Comment (carter-cundiff) [inline README.md]: will be placed to with the `containerize-dependencies` goal. -> will be copied to as part of the `containerize-dependencies` goal.
